PEP 709: Inlined comprehensions (#3029)
python/peps: pep-9999.rst

PEP: 9999
Title: Inlined comprehensions
Author: Carl Meyer <carl@oddbird.net>
Sponsor: Guido van Rossum <guido@python.org>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 24-Feb-2023
Python-Version: 3.12
Post-History:

Abstract
========

Comprehensions are currently compiled as nested functions, which provides
isolation of the comprehension's iteration variable, but is inefficient at
runtime. This PEP proposes to inline list, dictionary, and set comprehensions
into the function where they are defined, and provide the expected isolation by
pushing/popping clashing locals on the stack. This change makes comprehensions
much faster: up to 2x faster for a microbenchmark of a comprehension alone,
translating to 11% speedup for one sample benchmark derived from real-world code
that makes heavy use of comprehensions in the context of doing actual work.

> Motivation
> ==========

Reviewer: Not a huge deal, but FWIW, per PEP 1, PEP 12 and general convention, this section is IMO more Rationale than Motivation, as it mostly discusses the reasoning behind the proposed solution and the design decisions it makes rather than explains the motivation problem (performance, which is fairly self-evident).

```suggestion
Rationale
=========
```

             30 STORE_FAST               1 (x)
             32 RETURN_VALUE

There is no longer a separate code object, nor creation of a single-use function
object, nor any need to create and destroy a Python frame.

Isolation of the ``x`` iteration variable is achieved by the combination of the
new ``LOAD_FAST_AND_CLEAR`` opcode at offset ``6``, which saves any outer value
of ``x`` on the stack before running the comprehension, and ``30 STORE_FAST``,
which restores the outer value of ``x`` (if any) after running the
comprehension.

If the comprehension accesses variables from the outer scope, inlining avoids
the need to place these variables in a cell, allowing the comprehension (and all
other code in the outer function) to access them as normal fast locals instead.
This provides further performance gains.

This PEP only proposes to inline comprehensions occurring inside functions,
where fast-locals (``LOAD_FAST/STORE_FAST``) are used. Module-level
comprehensions will continue to create and call a function. Module-level
comprehensions are generally called only once (when the module is imported), so
optimizing the performance of calling them is unnecessary, and inlining them

> would require separate code paths in the compiler to handle a module global

Reviewer 2: It would also have backward-compatibility implications:

```
x = 3

def f():
    return x

print([f() for x in (1, 2)])
```

If we temporarily clear `x` while the module-level comprehension is ongoing, this will produce `(1, 2)` instead of `(3, 3)`.

namespace dictionary instead of fast-locals.

Backwards Compatibility
=======================

Comprehension inlining will cause the following visible behavior changes. No
changes in the standard library or test suite were necessary to adapt to these
changes in the implementation, suggesting the impact in user code is likely to
be minimal.

Specialized tools depending on undocumented details of compiler bytecode output
may of course be affected in ways beyond the below, but these tools already must
adapt to bytecode changes in each Python version.

locals() includes outer variables
---------------------------------

Calling ``locals()`` within a comprehension will include all locals of the
function containing the comprehension. E.g. given the following function::

  def f(lst):
      return [locals() for x in lst]

Calling ``f([1])`` in current Python will return::

  [{'.0': <list_iterator object at 0x7f8d37170460>, 'x': 1}]

> Where ``.0`` is the internal implementation detail of the synthetic sole

Reviewer: ```suggestion
where ``.0`` is the internal implementation detail of the synthetic sole
```

Nitpick, but this is a continuation of the previous sentance

> argument to the comprehension "function."

Reviewer: ```suggestion
argument to the comprehension "function".
```

Fix punctuation

Author: Interesting. I'd always been taught that periods and commas always go inside quotation marks. Searching it now I find that this is apparently American style, and British style puts them outside. (E.g. https://grammar.yourdictionary.com/grammar/punctuation/does-punctuation-go-inside-quotation-marks.html)

Do we prefer the British style in Python PEPs and docs?

Reviewer 2: I think a lot of American style now also puts punctuation outside of quotation marks, e.g. Wikipedia: https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style#Punctuation_inside_or_outside . I wouldn't want to impose this style if a PEP author prefers otherwise, but I do think it's more logical to put the punctuation outside the quotes.

Reviewer 3: AIUI it was once recommended to put punctuation inside quotes for the sake of movable type, but computer programmers rapidly settled on the form that has the accuracy needed for this industry. And since movable type isn't very relevant any more and computer programming kinda is, the "inside quotes" pattern is becoming the preferred one in most places.

Under this PEP, it will instead return::

  [{'lst': [1], 'x': 1}]

> Now including the outer ``lst`` variable as a local, and eliminating the

Reviewer: ```suggestion
This now includes the outer ``lst`` variable as a local, and eliminates the
```

I found the current wording a tad unclear at first, which this avoids

synthetic ``.0``.

> no comprehension frame in tracebacks

Reviewer: ```suggestion
No comprehension frame in tracebacks
```

At least sentence case in level 2+ headings

------------------------------------

Under this PEP, a comprehension will no longer have its own dedicated frame in a
stack trace. For example, given this function::

>   def f():
>     def g():
>       raise RuntimeError("boom")
> 
>     return [g() for x in [1]]

Reviewer: I found the structure here with the nested indent to be a little less clear and more distracting than two separate functions, when it doesn't appear to be needed to get the point across. Could it be simplified to

```suggestion
  def g():
      raise RuntimeError("boom")

  def f():
      return [g() for x in [1]]
```

instead (and normalized to standard 4-space indents used throughout the rest of the PEP)?

> Currently calling ``f()`` results in the following traceback:

Reviewer: ```suggestion
Currently, calling ``f()`` results in the following traceback:
```

Add comma and clarify the intended semantics slightly

.. code-block:: text

   Traceback (most recent call last):

>      File "<stdin>", line 1, in <module>
>      File "<stdin>", line 4, in f
>      File "<stdin>", line 4, in <listcomp>
>      File "<stdin>", line 3, in g

Reviewer: The extra line break makes this traceback not actually match the code snippit above. If you keep the current nested functions, this would in fact be

```suggestion
     File "<stdin>", line 1, in <module>
     File "<stdin>", line 5, in f
     File "<stdin>", line 5, in <listcomp>
     File "<stdin>", line 3, in g
```

However, if you apply my suggested change above, it would now be

```suggestion
     File "<stdin>", line 1, in <module>
     File "<stdin>", line 5, in f
     File "<stdin>", line 5, in <listcomp>
     File "<stdin>", line 2, in g
```

   RuntimeError: boom

Note the dedicated frame for ``<listcomp>``.

Under this PEP, the traceback looks like this instead:

.. code-block:: text

>    Traceback (most recent call last):
>      File "<stdin>", line 1, in <module>
>      File "<stdin>", line 4, in f
>      File "<stdin>", line 3, in g

Reviewer: Likewise, this should presumably be

```suggestion
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
     File "<stdin>", line 5, in f
     File "<stdin>", line 3, in g
```

or with the above change,

```suggestion
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
     File "<stdin>", line 5, in f
     File "<stdin>", line 2, in g
```

information.

It is theoretically possible that code using warnings with the ``stacklevel``
argument could observe a behavior change due to the frame stack change. In
practice, however, this seems unlikely. It would require a warning raised in
library code that is always called through a comprehension in that same
library, where the warning is using a ``stacklevel`` of 3+ to bypass the
comprehension and its containing function and point to a calling frame outside
the library. In such a scenario it would usually be simpler and more reliable
to raise the warning closer to the calling code and bypass fewer frames.

UnboundLocalError instead of NameError
--------------------------------------

Although the value of the comprehension iteration variable is saved and restored
to provide isolation, it still becomes a local variable of the outer function
under this PEP. Thus, if the comprehension iteration variable is accessed
outside the comprehension without ever being set outside the comprehension, as
in this function::

   def f(lst):
       items = [x for x in lst]
       return x

Under this PEP, calling ``f()`` will raise an ``UnboundLocalError``, whereas

> currently it raises a ``NameError``.

Reviewer: It could be noted that ``UnboundLocalError`` is a subclass of ``NameError``, so existing code will continue to work unchanged (unless it explicitly checks the exact exception type, which is usually a bad idea and presumably quite uncommon).

How to Teach This
=================

It is not intuitively obvious that comprehension syntax will or should result
in creation and call of a nested function. For new users not already accustomed
to the prior behavior, I suspect the new behavior in this PEP will be more
intuitive and require less explanation. ("Why is there a ``<listcomp>`` line in
my traceback when I didn't define any such function? What is this ``.0``
variable I see in ``locals()``?")

Reference Implementation
========================

> This PEP has a reference implementation in the form of `a PR against CPython main

Reviewer: ```suggestion
This PEP has a reference implementation in the form of `a PR against the CPython main
```

Missing definite article

branch <https://github.com/python/cpython/pull/101441>`_ which passes all tests.

The reference implementation performs the micro-benchmark ``./python -m pyperf

> timeit -s 'l = [1]' '[x for x in l]'`` 1.96x faster than ``main`` branch (in a

Reviewer 2: ```suggestion
timeit -s 'l = [1]' '[x for x in l]'`` 1.96x faster than the ``main`` branch (in a
```

build compiled with ``--enable-optimizations``.)

The reference implementation performs the ``comprehensions`` benchmark in the
`pyperformance <https://github.com/python/pyperformance>`_ benchmark suite
(which is not a micro-benchmark of comprehensions alone, but tests

> real-world-derived-code doing realistic work using comprehensions) 11% faster

Reviewer 2: ```suggestion
real-world-derived code doing realistic work using comprehensions) 11% faster
```

> than ``main`` branch (again in optimized builds.) Other benchmarks in

Reviewer: ```suggestion
than ``main`` branch (again in optimized builds). Other benchmarks in
```
